### ISSUE TITLE
Serious issue with unsafe operators

### ISSUE THREAD

**Issue by wbhart** [Nemocas/Nemo.jl#242]
The following is a serious bug that can hit us when using unsafe operators in combination with generics:

```
julia> using Nemo

Welcome to Nemo version 0.6.3

Nemo comes with absolutely no warranty whatsoever

julia> R, x = PolynomialRing(Nemo.ZZ, "x")
(Univariate Polynomial Ring in x over Integer Ring, x)

julia> S, y = PolynomialRing(R, "y")
(Univariate Polynomial Ring in y over Univariate Polynomial Ring in x over Integer Ring, y)

julia> C = y + 1
y+1

julia> B = y + 1
y+1

julia> C = Nemo.mul!(C, C, B)
y^2+2*y+1

julia> B
2*y+1
```
I believe the issue is caused because we don't make a deepcopy of generic objects when coercing them into their own parent ring.

We haven't noticed this issue before, because we are almost always using Flint objects instead of generic ones. For example, mul! on fmpz_poly's calls the Flint fmpz_poly_mul function and not the Julia generic code for mul!

The problem is, if two generic objects (e.g. polynomials) are different objects, but contain coefficients which are the same object, then modifying one can modify the other if it is the same Flint object, say.

I don't immediately see what the cheapest solution to this problem is. Just making deepcopies of everything in the unsafe operator functions kind of does away with the benefit of having them in the first place.

The idea of returning b when doing R(b), where R is the parent object for b, comes from Julia itself, which is the behaviour of BigInt, for example. But Julia is not implementing unsafe functions for BigInt's.

One solution might be to make a deepcopy of any object when coercing it into its parent ring. Another solution might be to ensure that no arithmetic operation can result in coefficients of the output === to coefficients of the input. That's a lot of code to check though. But maybe that is the most efficient solution.

**Comment by wbhart** [comment]
Actually, I think the possible solutions are clear:

1) If there is aliasing in the generic unsafe operators, then a deepcopy of *all* inputs must be made, because you cannot know if some part of the internals of one of the inputs aliases some part of one of the other inputs.

2) If there is aliasing in the C/Flint unsafe operators, a new output must be allocated and returned, because you cannot know if you are modifying a value which is unique.

Number 1 seems like the most logical from a performance perspective.

**Comment by wbhart** [comment]
We already make a copy of the input being aliased (just not the other inputs as well), in the case of generic unsafe operators. So we already have a performance hit in this case.

**Comment by wbhart** [comment]
Ok, I think I see what is happening now. The problem is with lines like:

```
setcoeff!(a, i - 1, coeff(b, i - 1))
```

These set the output coefficient of a polynomial to be the same object as an input object. Instead we need:

```
setcoeff!(a, i - 1, deepcopy(b, i - 1))
```

I'm unable to get a failure if I make this change, since output coefficients are always distinct from input coefficients and thus no two polynomials can share the same coefficients at any point.

We don't want to make setcoeff! do the deepcopy because things like:

```
setcoeff!(a, i - 1, b0*b1)
```
already create a new object when doing the multiplication b0*b1, so we don't want to make yet another copy of it.

We just have to be more careful when using setcoeff!!

**Comment by wbhart** [comment]
The other problem was that unsafe operators were setting the length of a polynomial without zeroing any additional coefficients. If those coefficients were ever used again, e.g. by addeq!, they'd have something already in them and the result would be wrong.

The unsafe operators should call set_length!, and this should zero any additional coeffs. The same probably applies to all the power series code.

**Comment by thofma** [comment]
I think we have some implicit aliasing rules, but they should be documented somewhere. As far as I understand, for a normal function (not an unsafe one), the output never shares memory/data with the input except in the following cases: `view`, `coeff` and `getindex` (for matrices). There everything is allowed.

I think this is our only rule when it comes to such matters.

**Comment by wbhart** [comment]
I have changed setcoeff! to take an optional flag. It now deepcopys the coefficient by default, but setting the flag to false suppresses the copy. Also, creating a matrix from an array now copies the array, since there is nothing stopping the user putting the same object in two arrays (it happened all through our test code!) And setindex! also needs to copy by default, though it is a pain in the neck to go through and replace A[i, j] = a*b with setindex!(A, i, j, a*b, false) all through our generic matrix code. I also found many cases where code that is supposed to work for PolyElems was just setting the coefficient array directly. Actually, I'm now buried about 6 layers deep in bugs related to our reliance on Flint.

**Comment by wbhart** [comment]
It's not possible for a normal function to have aliasing, since a normal function returns its outputs. They aren't passed to the function like in C. The issues here are all to do with two different object, e.g. two different polynomials, having the same objects inside them, e.g. the two different polynomials have a coefficient object in common. This is hard to prevent without making copies in lots of places. This will slow things down if we don't deal with it carefully.

**Comment by wbhart** [comment]
I have reconsidered this and rolled back the changes I made. This turned out to be a straight up bug in the + and - functions in the generic Poly module. The rule has always been that arithmetic operations must return new objects (with new contents). There were simply some deepcopys missing in parts of the + and - functions. 

It's perfectly fine for coefficients in polynomials or entries in matrices to alias one another. The rule for the unsafe operators is that they can only be used to modify objects that you control, that no one else contains a reference to (or to a part of).
